polymer_brushes/SCF.py

```diff
@@ -0,0 +1,899 @@
+    """Calculate Shultz-Zimm distribution from PDI and number average DP
+
+    Shultz-Zimm is a "realistic" distribution for linear polymers. Numerical
+    problems arise when the distribution gets too uniform, so if we find them,
+    default to an exact uniform calculation.
+    """
+    from scipy.special import gammaln
+
+    args = pdi, nn
+    if args in cache:
+        cache[args] = cache.pop(args)
+        return cache[args]
+
+    uniform = False
+
+    if pdi == 1.0:
+        uniform = True
+    elif pdi < 1.0:
+        raise ValueError("Invalid PDI")
+    else:
+        x = 1.0 / (pdi - 1.0)
+        # Calculate the distribution in chunks so we don't waste CPU time
+        chunk = 256
+        p_ni_list = []
+        pdi_underflow = False
+
+        for i in range(max(1, int((100 * nn) / chunk))):
+            ni = np.arange(chunk * i + 1, chunk * (i + 1) + 1, dtype=np.float64)
+            r = ni / nn
+            xr = x * r
+
+            p_ni = np.exp(np.log(x / ni) - gammaln(x + 1) + xr * (np.log(xr) / r - 1))
+
+            pdi_underflow = (p_ni >= 1.0).any()  # catch "too small PDI"
+            if pdi_underflow:
+                break  # and break out to uniform calculation
+
+            # Stop calculating when species account for less than 1ppm
+            keep = (r < 1.0) | (p_ni >= 1e-6)
+            if keep.all():
+                p_ni_list.append(p_ni)
+            else:
+                p_ni_list.append(p_ni[keep])
+                break
+        else:  # Belongs to the for loop. Executes if no break statement runs.
+            raise RuntimeError("SZdist overflow")
+
+    if uniform or pdi_underflow:
+        # NOTE: rounding here allows nn to be a double in the rest of the np.logic
+        p_ni = np.zeros(int(round(nn)))
+        p_ni[-1] = 1.0
+    else:
+        p_ni = np.concatenate(p_ni_list)
+        p_ni /= p_ni.sum()
+    cache[args] = p_ni
+
+    if len(cache) > 9000:
+        cache.popitem(last=False)
+
+    return p_ni
+
+
+def SCFprofile(
+    z,
+    chi=None,
+    chi_s=None,
+    h_dry=None,
+    l_lat=1,
+    mn=None,
+    m_lat=1,
+    phi_b=0,
+    pdi=1,
+    disp=False,
+    dist_model=SZdist,
+    clear_cache=False,
+):
+    """
+    Polymer end-tethered to an interface in a solvent.
+    Generate volume fraction profile for Refl1D based on real parameters.
+
+    The field theory is a lattice-based one, so we need to move between lattice
+    and real space. This is done using the parameters l_lat and m_lat, the
+    lattice size and the mass of a lattice segment, respectivley. We use h_dry
+    (dry thickness) as a convenient measure of surface coverage, along with mn
+    (number average molecular weight) as the real inputs.
+
+    Make sure your inputs for h_dry/l_lat and mn/m_lat match dimensions!
+    Angstroms and daltons are good choices.
+
+    Uses a numerical self-consistent field profile.\ [#Cosgrove]_\ [#deVos]_\ [#Sheridan]_
+
+    **Parameters**
+        *chi*
```

Comment:
Use numpydoc type formatting, e.g.

```
Parameters
-----------
```

Unless, this docstring is taken from refl1d, in which case leave as-is.